generator/src/main/java/com/graphicsfuzz/generator/fuzzer/OpaqueExpressionGenerator.java

```diff
@@ -802,4 +804,60 @@ public Expr apply(Expr expr, BasicType type, boolean constContext, int depth,
 
   }
 
+  private class IdentityRewriteComposite extends AbstractIdentityTransformation {
+    private IdentityRewriteComposite() {
+      // all non-boolean vector/matrix types
+      super(BasicType.allNumericTypes().stream().filter(
+          item -> !BasicType.allScalarTypes().contains(item)).collect(Collectors.toList()),
+          false);
+    }
+
+    @Override
+    public Expr apply(Expr expr, BasicType type, boolean constContext, int depth,
+                      Fuzzer fuzzer) {
+      // v -> (true ? vecX(..., identity(v[Y]), ...) : _)
+      // v -> (false ? _ : vecX(..., identity(v[Y]), ...))
+      // where v is a vector of size X, y is the random entry in v we want to apply
+      // identities to, and ... is the other entries in v that we don't change.
+      // Similarly for matrices.
+
+      assert BasicType.allVectorTypes().contains(type)
+          || BasicType.allMatrixTypes().contains(type);
+
+      final int numColumns =
+          (BasicType.allVectorTypes().contains(type)
+          ? type.getNumElements() : type.getNumColumns());
+      final int indexToApplyIdentities = generator.nextInt(numColumns);
+      final List<Expr> nonscalarEntryList = new ArrayList<>();
+      for (int i = 0; i < numColumns; i++) {
+        // v + vec2(0.0)[0]; is not what we want to do, so we wrap expr into
+        // parentheses for (v + vec2(0.0))[0]
+        if (i == indexToApplyIdentities) {
+          nonscalarEntryList.add(applyIdentityFunction(
+              new ArrayIndexExpr(new ParenExpr(expr.clone()),
+                  new IntConstantExpr(String.valueOf(i))),
+              type.getElementType(),
+              constContext, depth, fuzzer));
+        } else {
+          nonscalarEntryList.add(new ArrayIndexExpr(new ParenExpr(expr.clone()),
+              new IntConstantExpr(String.valueOf(i))));
+        }
+      }
+      return identityConstructor(expr,
+          new TypeConstructorExpr(type.toString(), nonscalarEntryList));
+    }
+
+    @Override
+    public boolean preconditionHolds(Expr expr, BasicType basicType) {
+      if (!super.preconditionHolds(expr, basicType)) {
+        return false;
+      }
+      if (!(expr instanceof VariableIdentifierExpr)) {
+        // limit blowup of identities.
+        return false;
+      }
+      return SideEffectChecker.isSideEffectFree(expr, shadingLanguageVersion);
```

Comment:
^ super.preconditionHolds(...) does the side-effect checking.  So it is redundant and confusing to do it again here.

I would thus change the whole method body to:

return super.preconditionHolds(expr, basicType) && expr instanceof VariableIdentifierExpr;